### PR TITLE
[java] Support Java 19 (pmd7)

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

Merge of #4038 into pmd/7.0.x

### CODE REVIEW

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/internal/LanguageLevelChecker.java [hunk 1]
```diff
@@ -221,6 +242,11 @@ private enum Keywords implements LanguageFeature {
          */
         PERMITS_AS_A_TYPE_NAME(15, "permits"),
 
+        /**
+         * ContextualKeyword since Java 19 Preview.
+         */
+        WHEN_AS_A_TYPE_NAME(19, "when"),
+
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/internal/LanguageLevelChecker.java:249]: I'm not sure, if this is correct... You *can* use `when` as a class name: `public class when {}` compiles fine with java 19 preview. Having this in the "Keywords" enum, we would prevent this, right? In that case, we probably should remove it...

File: pmd-java/etc/grammar/Java.jjt [hunk 2]
```diff
@@ -1712,7 +1719,9 @@ void EqualityExpression() #void:
 void Pattern() #void:
 {}
 {
-    PrimaryPattern() [ GuardedPatternCondition() #GuardedPattern(2) ]
+    LOOKAHEAD(ReferenceType() "(") RecordPattern()
+    | LOOKAHEAD("(") ParenthesizedPattern()
+    | TypePattern() [ LOOKAHEAD({getToken(1).kind == SC_AND && jdkVersion < 19}) GuardedPatternCondition() #GuardedPattern(2) ]
```
Comment (adangel) [inline pmd-java/etc/grammar/Java.jjt:1724]: This lookahead including the GuardedPatternCondition can hopefully be removed with one of the later java versions. Right now, it's only needed for java 18 preview - which parses the source really different...

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 28 violations,<br />
introduces 23 new violations, 0 new errors and 0 new configuration errors,<br />
removes 26 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/ed9235759816418798c86e479956c741/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 57095 violations,<br />
introduces 34417 new violations, 2 new errors and 0 new configuration errors,<br />
removes 138133 violations, 25 errors and 6 configuration errors.<br />
<a href="https://chunk.io/pmd/ed9235759816418798c86e479956c741/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/internal/LanguageLevelChecker.java (hunk 1 above)
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/internal/LanguageLevelChecker.java:249]: If we can use when as a type name then I suppose you're right, the only thing we need to add in this enum is the `CASE_REFINEMENT` above

Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/internal/LanguageLevelChecker.java:249]: I'll remove it then

**Pull request merged**